File: src/main/java/io/felipeandrade/gsw2/block/GSWBlock.java
package io.felipeandrade.gsw2.block;

import net.minecraft.block.Block;
import net.minecraft.item.BlockItem;
import net.minecraft.item.Item;
import net.minecraft.item.ItemGroup;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;

import io.felipeandrade.gsw2.common.Localizeable;
import io.felipeandrade.gsw2.common.Registrable;

import static io.felipeandrade.gsw2.GSW2Mod.MOD_ID;

public class GSWBlock extends Block implements Registrable, Localizeable {

    protected String unlocalizedName;
    protected ItemGroup itemGroup;

    public GSWBlock(String unlocalizedName, ItemGroup itemGroup, Settings settings) {
        super(settings);
        this.unlocalizedName = unlocalizedName;
        this.itemGroup = itemGroup;
    }

    @Override
    public void register() {
        Registry.register(Registry.BLOCK, new Identifier(MOD_ID, unlocalizedName), this);
        Registry.register(Registry.ITEM, new Identifier(MOD_ID, unlocalizedName), new BlockItem(this, new Item.Settings().group(itemGroup)));
    }

    @Override
    public String unlocalizedName() {
        return unlocalizedName;
    }

}

File: gsw-mod/src/main/kotlin/io/felipeandrade/gsw/block/GSWBlock.kt
package io.felipeandrade.gsw.block

import io.felipeandrade.gsw.GSWMod
import io.felipeandrade.gsw.common.Registrable
import net.minecraft.block.Block
import net.minecraft.item.BlockItem
import net.minecraft.item.Item
import net.minecraft.item.ItemGroup
import net.minecraft.registry.Registries
import net.minecraft.registry.Registry
import net.minecraft.util.Identifier

open class GSWBlock(
    val unlocalizedName: String,
    var itemGroup: ItemGroup,
    settings: Settings
) : Block(settings), Registrable {
    override fun register() {
        Registry.register(Registries.BLOCK, Identifier(GSWMod.MOD_ID, unlocalizedName), this)
        Registry.register(
            Registries.ITEM,
            Identifier(GSWMod.MOD_ID, unlocalizedName),
            BlockItem(this, Item.Settings())
        )
    }
}